### PR TITLE
build: disable RocksDB assertions, except on race builds

### LINKED ISSUE #15604: engine: TestConcurrentBatch is flaky (locally)
`TestConcurrentBatch` has recently (in the last couple of weeks, I think) become flaky on my laptop. There are a bunch of slow batches, each taking longer than the last (this is a partial log):

```
W170502 12:50:40.391933 293 storage/engine/rocksdb.go:1220  batch [355017/4194314/0] commit took 12.411354136s (>500ms):
goroutine 293 [running]:
runtime/debug.Stack(0x4abc360, 0x4ff5a40, 0xc420014e80)
	/opt/local/lib/go/src/runtime/debug/stack.go:24 +0x79
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).commitInternal(0xc4203e0480, 0x4030300, 0x4be0868, 0xc42002cc00)
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1220 +0x14d
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).Commit(0xc4203e0480, 0xc406591100, 0x4670001, 0xc42041bfc8)
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1189 +0x704
github.com/cockroachdb/cockroach/pkg/storage/engine.TestConcurrentBatch.func2(0xc420346300, 0x4ffa7a0, 0xc4203e0480)
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb_test.go:391 +0x36
created by github.com/cockroachdb/cockroach/pkg/storage/engine.TestConcurrentBatch
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb_test.go:392 +0x75c
W170502 12:50:41.449260 305 storage/engine/rocksdb.go:1220  batch [355017/4194314/0] commit took 13.480834217s (>500ms):
goroutine 305 [running]:
runtime/debug.Stack(0x4abc360, 0x4ff5a40, 0xc420014e80)
	/opt/local/lib/go/src/runtime/debug/stack.go:24 +0x79
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).commitInternal(0xc4200b8900, 0x0, 0x0, 0x0)
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1220 +0x14d
github.com/cockroachdb/cockroach/pkg/storage/engine.(*rocksDBBatch).Commit(0xc4200b8900, 0x0, 0x0, 0x0)
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:1189 +0x704
github.com/cockroachdb/cockroach/pkg/storage/engine.TestConcurrentBatch.func2(0xc420346300, 0x4ffa7a0, 0xc4200b8900)
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb_test.go:391 +0x36
created by github.com/cockroachdb/cockroach/pkg/storage/engine.TestConcurrentBatch
	/Users/bdarnell/work/cockroach/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb_test.go:392 +0x75c
I170502 12:50:41.449954 211 storage/engine/rocksdb.go:447  closing rocksdb instance at "/var/folders/xw/y575gcfd7wx30kskwzd6p2qr0000gn/T/TestConcurrentBatch351385987"
--- FAIL: TestConcurrentBatch (16.33s)
	rocksdb_test.go:416: write took 13.4s
```

When the test passes, it takes about 4 seconds on this laptop (a macbook pro, with filevault enabled). 

I think this is probably a matter of other stuff going on on the machine. It fails a large fraction of the time when I run `make test` and there is stuff to rebuild, but it passes reliably when run in isolation or when there is no rebuilding going on. So it's not really a problem with the test per se, but I'd like to see it be more robust against concurrent activity (or for us to figure out a way for `make test` to not cause so much concurrent I/O)

### PR DESCRIPTION
**Pull request by benesch**
@tamird, passing `-DCMAKE_BUILD_TYPE` opened a can of worms. I think we should hold off on passing `-CMAKE_BUILD_TYPE` until facebook/rocksdb#2300 or its equivalent lands. In the meantime, this patch is nice and focused and hopefully easy to cherry pick to the 1.0 branch.

---

In the transition to the new Make-based build system for C dependencies (4d72e12), RocksDB assertions were inadvertently enabled for all builds. Restore the old behavior, which only enabled assertions for race builds (cockroachdb/c-rocksdb#27).

Fixes #15604.

### CODE REVIEW

File: build/common.mk [hunk 1]
```diff
@@ -210,6 +210,9 @@ CMAKE_FLAGS := $(if $(MINGW),-G 'MSYS Makefiles')
 override USE_STDMALLOC := $(findstring stdmalloc,$(TAGS))
 STDMALLOC_SUFFIX := $(if $(USE_STDMALLOC),_stdmalloc)
 
+ENABLE_ROCKSDB_ASSERTIONS := $(findstring race,$(TAGS))
+ENABLE_ROCKSDB_ASSERTIONS := $(if $(ENABLE_ROCKSDB_ASSERTIONS),_assert)
```
Comment (tamird) [inline build/common.mk]: why the intermediate value rather than
```
ENABLE_ROCKSDB_ASSERTIONS := $(if $(findstring race,$(TAGS)),_assert)
```

**Review by tamird — APPROVED** [review]

**Comment by benesch** [comment]
I should lose coding privileges at this late hour. Take another look?

**Comment by tamird** [comment]
Ha, my review wasn't spot on either. :lgtm: and :pray: :moon: 

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15943#-:-KkE5Cyn5n_GmVKlEon3:b-81aboe)*

**Comment by benesch** [comment]
👍 🛏

**Review by petermattis — APPROVED** [review]
LGTM

**Comment by benesch** [comment]
TFTRs!

**Comment by mberhault** [comment]
this conflicts with #15719 and #15753. I don't mind pulling in build changes, but someone should check if we want those or rewrite the patch.

**Comment by petermattis** [comment]
#15719 and #15753 are not really important for 1.0.1. This change is.

**Comment by benesch** [comment]
I'll rewrite the patch!

**Pull request merged**